### DISCUSSION TITLE
Performance problem with `verifyPublicKeyParameterIntegrity`

### DISCUSSION THREAD

**Discussion by bjansen** [pgpainless/pgpainless#349] [Q&A]
Hi,

I'm working on an application which sends and receives lots of PGP-encrypted files. I'm trying to cut down our build times, so I used YourKit to profile our integration tests and apparently a huge amount of time is spent in `BigInteger.modPow` (41%!):

![image](https://user-images.githubusercontent.com/281528/207586952-eb9fe437-3a8d-4acf-8f03-8633ee52cae3.png)

Roughly 50% of those 41% seem to be spent in `PublicKeyParameterValidationUtil.verifyCanDecrypt` and `PublicKeyParameterValidationUtil.verifyCanSign`, which means around 20% of the total integration tests time is spent verifying PGP keys. This seems a lot. FTR, our tests currently run in around 12-13 minutes.

I was wondering if there was a way to skip `PublicKeyParameterValidationUtil.verifyPublicKeyParameterIntegrity` under certain circumstances, like test executions. Or maybe only call it once per `PGPSecretKey`?

**Comment by vanitasvitae** [comment]
There is no way to do that out of the box currently.
Now that I see how great the performance impact really is though, I believe it makes sense to employ some configuration option to skip these checks (which are only useful in certain scenarios anyways, see https://www.kopenpgp.com/), or alternatively cache validation results to skip already validated keys.

Thanks for reporting!

**Reply by bjansen** [reply]
My use-case might be specific or my benchmark wrong though. I'd be interested to see the results of a separate benchmark on another project. Maybe profiling pgpainless' own unit tests would yield a different result?

**Reply by vanitasvitae** [reply]
I guess your test suite entails using lots of ElGamal / DSA secret keys?

Yourkit seems like a useful tool. I will see if they can supply me with an Opensource license of their tool. Thanks for the tipp!

**Reply by bjansen** [reply]
Actually no, they're RSA 4096 keys.
And yes, YourKit is incredibly useful to analyze performance problems :)

**Reply by vanitasvitae** [reply]
Aaah, sorry I just glanced over the `PublicKeyParameterValidationUtil` class, looking for `BigInteger.modPow()` method calls, so I missed the `verifyCanSign()/verifyCanEncrypt()` calls from your screenshot.

The validation code for RSA signing keys creates and verifies a test signature to check if the key is usable.
With 1.4, those checks will be disabled by default.

As a user of the library, would you like to be able to control those checks more fine-grained? I could for example think of a configuration API which could allow users to enable/disable single checks.

Also, a caching system which prevents repeated checks on the same key sounds like a cool project too, but I'm not sure yet, how to properly implement that.

**Reply by bjansen** [reply]
I wasn't even aware that key overwriting was a thing before reading the `PublicKeyParameterValidationUtil` class. I'm not sure why users would want to `verifyCanSign()` but not `verifyCanEncrypt()`, so I guess you don't need to implement a fine-grained API.

**Comment by vanitasvitae** [comment]
I decided for 1.4 to disable the countermeasures against KOpenPGP.
They can however be enabled if required by calling `Policy.setEnableKeyParameterValidation(true)`.

**Comment by bjansen** [comment]
I think I managed to bypass the verifications using Mockito:

```java
Mockito.mockStatic(PublicKeyParameterValidationUtil.class);
```

`modPow` is still a hot spot, but now it's only used for actual encryption/signature of files:

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/281528/208053814-9ffae89e-4d63-44b3-bf07-14a65e60fb73.png">

Build time is a little better on my machine (macbook pro 13), but somehow it doesn't look like it makes much of a difference on our CI (a Jenkins agent running Debian IIRC). 

I wonder if the performance of `BigInteger` depends on the JVM, the OS, or maybe CPU features.
